cpp/benchmarks/common/random_distribution_factory.hpp

```diff
@@ -62,66 +74,80 @@ double geometric_dist_p(T range_size)
+    : lower_bound(lower_bound), upper_bound(upper_bound), engine(engine), dist(gen)
+  {
+  }
 
-template <typename T, std::enable_if_t<cudf::is_floating_point<T>()>* = nullptr>
-auto make_geometric_dist(T range_start, T range_end)
-{
-  long double const range_size = range_end - range_start;
-  return std::exponential_distribution<T>(geometric_dist_p(range_size));
-}
+  __device__ T operator()(size_t n)
+  {
+    engine.discard(n);
+    return dist(engine) + lower_bound;
+  }
 
-template <typename T>
-using distribution_fn = std::function<T(std::mt19937&)>;
+  T lower_bound;
+  T upper_bound;
+  thrust::minstd_rand engine;
+  Generator dist;
+};
 
-template <typename T, std::enable_if_t<cuda::std::is_integral_v<T>, T>* = nullptr>
-distribution_fn<T> make_distribution(distribution_id did, T lower_bound, T upper_bound)
-{
-  switch (did) {
-    case distribution_id::NORMAL:
-      return [lower_bound, dist = make_normal_dist(upper_bound - lower_bound)](
-               std::mt19937& engine) mutable -> T { return dist(engine) + lower_bound; };
-    case distribution_id::UNIFORM:
-      return [dist = make_uniform_dist(lower_bound, upper_bound)](
-               std::mt19937& engine) mutable -> T { return dist(engine); };
-    case distribution_id::GEOMETRIC:
-      return [lower_bound, upper_bound, dist = make_geometric_dist(lower_bound, upper_bound)](
-               std::mt19937& engine) mutable -> T {
-        if (lower_bound <= upper_bound)
-          return dist(engine);
-        else
-          return lower_bound - dist(engine) + lower_bound;
-      };
-    default: CUDF_FAIL("Unsupported probability distribution");
+template <typename T, typename Generator>
+struct abs_value_generator : value_generator<T, Generator> {
+  using super_t = value_generator<T, Generator>;
+  abs_value_generator(T lower_bound, T upper_bound, thrust::minstd_rand& engine, Generator gen)
+    : super_t{lower_bound, upper_bound, engine, gen}
+  {
   }
-}
 
-template <typename T, std::enable_if_t<cudf::is_floating_point<T>()>* = nullptr>
+  __device__ T operator()(size_t n)
+  {
+    super_t::engine.discard(n);
+    return abs(super_t::dist(super_t::engine)) + super_t::lower_bound;
+  }
+};
+
+template <typename T>
+using distribution_fn = std::function<rmm::device_uvector<T>(thrust::minstd_rand&, size_t)>;
+
+template <
+  typename T,
+  std::enable_if_t<cuda::std::is_integral_v<T> or cuda::std::is_floating_point_v<T>, T>* = nullptr>
 distribution_fn<T> make_distribution(distribution_id dist_id, T lower_bound, T upper_bound)
 {
   switch (dist_id) {
     case distribution_id::NORMAL:
-      return [lower_bound, dist = make_normal_dist(upper_bound - lower_bound)](
-               std::mt19937& engine) mutable -> T { return dist(engine) + lower_bound; };
+      return [lower_bound, upper_bound, dist = make_normal_dist(upper_bound - lower_bound)](
+               thrust::minstd_rand& engine, size_t size) mutable -> rmm::device_uvector<T> {
```

Comment:
Why do these lambdas all need to be mutable? I asked this elsewhere too, so I'm guessing that `thrust::tabulate` or the `value_generator` requires a mutable parameter somewhere.

Reply:
Not sure why previous code was `mutable`. Only `engine` is modified in this lambda, which is a reference parameter.

Reply:
updated.

Reply:
My guess is that it needed to be `mutable` to update the element index, in the implementation that did not use `tabulate`. I probably switched to `tabulate` and forgot to remove `mutable`.